### PR TITLE
api: document the runtime guarding for API behavior change

### PR DESCRIPTION
**Pull request by soulxu**
Signed-off-by: He Jie Xu <hejie.xu@intel.com>

Commit Message: api: document the runtime guarding for API behavior change
Additional Description:
Risk Level: low
Testing: n/a
Docs Changes: API doc

### CODE REVIEW

**Comment by soulxu** [comment]
cc @mattklein123 @htuch @alyssawilk @envoyproxy/api-shepherds

**Comment by wbpcode** [comment]
docs only update.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -168,9 +169,18 @@ Some changes in Envoy are deemed worthy of runtime guarding. Instead of just rep
...
-refactors such as replacing Envoy's buffer implementation) and most user-visible
+refactors such as replacing Envoy's buffer implementation), most user-visible
 non-config-guarded changes to protocol processing (for example additions or changes to HTTP headers or
-how HTTP is serialized out) for non-alpha features. Feel free to tag @envoyproxy/maintainers
+how HTTP is serialized out) for non-alpha features and fixing API behavior.
+
+When using runtime guarding for API behavior change, follow the rules below:
+* If the API works without any meaningful behavior, it can be fixed directly without any runtime guarding.
+* If the API works without any meaningful behavior but the fixing is breaking change (eg. API returns a rejection),
+  the rejection should be added with runtime guarding.
```
Comment (soulxu) [inline CONTRIBUTING.md:179]: s/the rejection should be added with runtime guarding/it should be added with runtime guarding/

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -113,7 +113,8 @@ versioning guidelines:
   inline with protos as per the [API contribution guidelines](api/CONTRIBUTING.md). If a change applies
   to multiple sections of the release notes, it should be noted in the first (most important) section
   that applies. For instance, a bug fix that introduces incompatible behavior should be noted in
-  `Incompatible Behavior Changes` but not in `Bug Fixes`.
+  `Incompatible Behavior Changes` but not in `Bug Fixes`. Also should follow the runtime guarding rules
```
Comment (alyssawilk) [inline CONTRIBUTING.md:116]: Also -> It should also

**Review by alyssawilk — COMMENTED** [review]
cool thanks for trying to improve docs here.  Have some clarifying questions if you don't mind?
/wait

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -168,9 +169,18 @@ Some changes in Envoy are deemed worthy of runtime guarding. Instead of just rep
 old code with new code, both code paths are supported for between one Envoy release (if it is
 guarded due to performance concerns) and a full deprecation cycle (if it is a high risk behavioral
 change). Generally as a community we try to guard both high risk changes (major
-refactors such as replacing Envoy's buffer implementation) and most user-visible
+refactors such as replacing Envoy's buffer implementation), most user-visible
 non-config-guarded changes to protocol processing (for example additions or changes to HTTP headers or
-how HTTP is serialized out) for non-alpha features. Feel free to tag @envoyproxy/maintainers
+how HTTP is serialized out) for non-alpha features and fixing API behavior.
+
+When using runtime guarding for API behavior change, follow the rules below:
```
[the remark this answers is not in the document]
Comment (htuch) [reply CONTRIBUTING.md:176]: Yeah, I'm not sure exactly what we want to capture here. I know this follows from https://github.com/envoyproxy/envoy/pull/24081#discussion_r1027255491, where I think the intent is to say "if we made some giant mistake an didn't actually implement the documented API, we can fix the implementation to get it to match this with runtime guards". But I might not have fully captured the range of intended conditions in that.

File: CONTRIBUTING.md (hunk 2 above)
Comment (soulxu) [reply CONTRIBUTING.md:116]: got it, thanks!

File: CONTRIBUTING.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (alyssawilk) [reply CONTRIBUTING.md:176]: Does this just fall under behavioral changes?  Any major behavioral change should be guarded, whether it's behind a config knob or not.

[the remark this answers is not in the document]
Comment (soulxu) [reply CONTRIBUTING.md:176]: > Does this just fall under behavioral changes? 

Yes, it falls under behavioral changes. I guess behavioral changes are broader, like we could change the HTTP protocol behavior. 

> Any major behavioral change should be guarded, whether it's behind a config knob or not.

emm...not sure I understand the question, the `config knob`, do you mean the API field we defined? If yes, that is probably the API behavior that I said, the behavior when the user changes config.

**Comment by soulxu** [comment]
This is probably not having enough thought and providing enough meaning explanation. Let me close this PR. Thanks for all the review and feedback. @alyssawilk @htuch

**Pull request closed without merging**